src/runtime/HalideBuffer.h

```diff
@@ -167,20 +168,47 @@ class Buffer {
         if (alloc) {
             int new_count = --(alloc->ref_count);
             if (new_count == 0) {
+                if (buf.dev == 0) {
+                    // Adopt any orphan device allocation so that we can kill it.
+                    buf.dev = alloc->orphan_dev;
+                }
```

Comment:
For clarity, how about setting alloc->orphan_dev to 0 here, as it is technically no longer an orphan. The assert below also becomes simpler if you do this.